### ISSUE TITLE
Use Eigen SparseCholesky for DCP checking with sparse problem data

### ISSUE THREAD

**Issue by clazarus** [cvxpy/cvxpy#1459]
**Describe the bug**
When attempting to minimize a quadratic form (similar to the factor covariance model found at https://colab.research.google.com/github/cvxgrp/cvx_short_course/blob/master/applications/portfolio_optimization.ipynb) with a large (~18k entries) sparse diagonal matrix I run into an extremely slow (+10 minutes) compilation time when using MOSEK.

Below I include minimal code to reproduce the issue. Notice that `obj_func` and `obj_func2` are mathematically equivalent. `obj_func` compiles quickly with both MOSEK and OSQP. Whereas `obj_func2` compiles quickly with OSQP but stalls with MOSEK.

**To Reproduce**

```
import numpy as np
import cvxpy as cp
from scipy import sparse as sp

n = 18000
m = 80
np.random.seed(42)
sigma_tilde = np.random.randn(m, m)
sigma_tilde = sigma_tilde.T.dot(sigma_tilde)
D = sp.diags(np.random.uniform(0, 0.9, size=n))
F = np.random.randn(n, m)
w = cp.Variable(n)
f = cp.Variable(m)

obj_func = cp.quad_form(f, sigma_tilde) + cp.sum_squares(np.sqrt(D)@(w))
obj_func2 = cp.quad_form(f, sigma_tilde) + cp.quad_form(w, D)

prob = cp.Problem(cp.Minimize(obj_func2), 
                     [cp.sum(w) == 1,
                      f == F.T@w,
                      0 <= w,
                      w <= 1])

#prob.solve(verbose=True, solver='OSQP')
# Using MOSEK leads to a +9 minute compilation time
prob.solve(verbose=True, solver='MOSEK')
```

**Expected behavior**
Fast (under a few seconds) compilation.

**Output**
```
===============================================================================
                                     CVXPY                                     
                                    v1.1.13                                    
===============================================================================
(CVXPY) Aug 04 05:08:21 PM: Your problem has 18080 variables, 4 constraints, and 0 parameters.
(CVXPY) Aug 04 05:08:21 PM: It is compliant with the following grammars: DCP, DQCP
(CVXPY) Aug 04 05:08:21 PM: (If you need to solve this problem multiple times, but with different data, consider using parameters.)
(CVXPY) Aug 04 05:08:21 PM: CVXPY will first compile your problem; then, it will invoke a numerical solver to obtain a solution.
-------------------------------------------------------------------------------
                                  Compilation                                  
-------------------------------------------------------------------------------
(CVXPY) Aug 04 05:08:21 PM: Compiling problem (target solver=MOSEK).
(CVXPY) Aug 04 05:08:21 PM: Reduction chain: Dcp2Cone -> CvxAttr2Constr -> ConeMatrixStuffing -> MOSEK
(CVXPY) Aug 04 05:08:21 PM: Applying reduction Dcp2Cone
```
Process remains at this step for about 10 minutes and all the cores in my machine are maxed out.
**Version**
 - OS: macOS Big Sur Version 11.3.1 (20E241)
 - CVXPY Version: 1.1.13

**Additional context**
My intuition is that the issue is related to the internal representation of the problem and a specific step in the reduction chain. An equivalent problem (included above) is solved in under a second and indicates that the reduction chain "path" leads to a different representation which perhaps avoids a dense representation of the sparse matrix.

**Comment by SteveDiamond** [comment]
The problem is line 192 of quad_form.py, in decom_quad. We need the factorization to work for sparse matrices (without conversion to dense).

**Comment by rileyjmurray** [comment]
@SteveDiamond we could check if MOSEK is installed, and if so, use https://docs.mosek.com/latest/pythonapi/optimizer-env.html#mosek.env.computesparsecholesky to compute the sparse Cholesky decomposition. I also looked at SciPy but it seems they don't support sparse Cholesky factorization. We can determine if we need Cholesky for ``A`` or ``-A`` (posdef and negdef respectively) by looking at any nonzero diagonal entry of ``A`` (a matrix ``A`` is positive definite only if its diagonal is elementwise positive).

**Comment by aszekMosek** [comment]
In the first model it only needs to factorize of a small matrix, and in the second a huge matrix. So the first one is closer to the direct conic model of

https://docs.mosek.com/modeling-cookbook/qcqo.html#example-factor-model

and goes to show that if you have the factor model and a conic solver then it pays off to work directly in conic form, rather than spending time to form the QP just to have someone (CVXPY of MOSEK internally) spend even more time later undoing that work by computing Cholesky.

Apologies for a slight offtopic.

**Comment by SteveDiamond** [comment]
@aszekMosek agreed! In practice we solved this issue with a reformulation. It would be better though if the naive quadratic formulation also worked well.

**Comment by rileyjmurray** [comment]
Follow-up. A few comments back I suggested we use sparse Cholesky to determine positive definiteness, but pointed out that SciPy has no such functionality. We could consider using [Eigen's SparseCholesky module](https://eigen.tuxfamily.org/dox/group__SparseCholesky__Module.html). We already rely on Eigen for cvxcore, so this doesn't add a dependency. 

@akshayka @SteveDiamond what do you think? If using some of Eigen's sparse linear algebra functionality is on the table, then we could even use it as part of canonicalization, rather than just checking if something is DCP.

**Comment by SteveDiamond** [comment]
I'd be open to using Eigen in this way.

**Comment by rileyjmurray** [comment]
Resolved by #2125! To appear in CVXPY 1.4.
